README.MD
# Zephyr BLE sensor application
This application is a BLE sensor application for the Zephyr RTOS. It is designed to send advertising packets with sensor data.

## Environment
- nRF connect SDK v2.4.0
- [Seeed Xiao BLE sense](https://wiki.seeedstudio.com/XIAO_BLE/)

## Flash
🔗 [Reference](https://medium.com/@aloysiousbenoy/flash-nrf52840-using-uf2-no-more-copy-paste-75044b2160a7)

## Default sensor configuration
- Default accelerometer full-scale range: +/- 4g
- Default gyroscope full-scale range: +/- 500 dps
- Default accelerometer output data: 104 Hz
- Default gyroscope output data: 104 Hz

## BLE advertising name format
IMUXXXXXXXXXXXX 
  > with "XXXXXXXXXXXX" is the 6B device MAC address

## BLE indication message format
|  | **Frame Counter** | **ACL X** | **ACL Y** | **ACL Z** | **MAG X** | **MAG Y** | **MAG Z** |
|  --- | --- | --- | --- | --- | --- | --- | --- |
| **Length** | 4 | 4 | 4 | 4 | 4 | 4 | 4 |
|  |  |  |  |  |  |  |  |
> **Frame Counter**: 32-bit unsigned integer\
> **ACL**: 32-bit float\
> **MAG**: 32-bit float

CHANGELOG.md
## Version 0.0.1 (Oct 10, 2023)
- Initial release
- BLE operation as a broadcaster
- Sensor data is sent via BLE advertising packets

## Version 0.0.2 (Oct 19, 2023)
- Change BLE operation from broadcaster to connectable peripheral
- Sensor data is sent via indication messages
- Including device MAC address in advertising field "IMUXXXXXXXXXXXX"
- Reserved "COMMAND characteristic"

## Version 0.0.3 (Nov 10, 2023)
- Fix BLE advertising after connection
- Enable auto trigger connection parameter update
- Enable power management

src/main.c
/*******************************************************************************
* Title                 :    
* Filename              :   main.c
* Author                :   ItachiVN
* Origin Date           :   2023/10/04
* Version               :   0.0.0
* Compiler              :   nRF connect SDK v2.4
* Target                :   nRF52840 SoC
* Notes                 :   None
*******************************************************************************/


/******************************************************************************
* Includes
*******************************************************************************/
#include <zephyr/kernel.h>

#include "ble\ble_gatt.h"
#include "ble\bluetoothle.h"
/******************************************************************************
* Module Preprocessor Constants
*******************************************************************************/
#define FW_MAJOR_VERSION                (0)
#define FW_MINOR_VERSION                (0)
#define FW_BUIILD_VERSION               (3)

#define SAMPLING_RATE_MS                (25)
#define SENSOR_PACKET_LEN               (24) // 4B for each axis * 6 axis
#define SENSOR_DATA_HEADER              (0xAA)

/******************************************************************************
* Module Preprocessor Macros
*******************************************************************************/
#include "zephyr/logging/log.h"
#define MODULE_NAME			        main
#define MODULE_LOG_LEVEL	        LOG_LEVEL_WRN
LOG_MODULE_REGISTER(MODULE_NAME, MODULE_LOG_LEVEL);

/******************************************************************************
* Module Typedefs
*******************************************************************************/

/******************************************************************************
* Module Variable Definitions
*******************************************************************************/
volatile static bool is_ble_connect=false;

void on_ble_connect(void)
{
    is_ble_connect = true;
    LOG_INF("BLE connected");
    if (ble_adv_stop() != 0)
    {
        LOG_ERR("BLE adv stop failed");
    }
}

void on_ble_disconnect(void)
{
    is_ble_connect = false;
    LOG_INF("BLE disconnected");
    if(ble_adv_start() != 0)
    {
        LOG_ERR("BLE adv start failed");
    }
}

int on_cmd_write_cb(void* p_data, void* p_len)
{
    // Dump receive data
    LOG_INF("Command received, recv %dB", *(int*)p_len);
    LOG_HEXDUMP_INF(p_data, *(int*)p_len, "raw_payload: ");
    return 0;
}

void* on_cmd_read_cb(void* p_data, void* p_len)
{
    static char default_char_value[]="CMD CHAR READ CB TRIGGER";
    *(int*)p_len = strlen(default_char_value);
    return default_char_value;
}


ble_callback_t ble_callback = 
{
    .ble_connected_cb = &on_ble_connect,
    .ble_disconnected_cb = &on_ble_disconnect,
};

ble_custom_gatt_cb_t ble_custom_gatt_cb = 
{
    .custom_char_read_cb = &on_cmd_read_cb,
    .custom_char_write_cb = &on_cmd_write_cb,
};
/******************************************************************************
* Function Prototypes
*******************************************************************************/
int ble_app_init(void);
/******************************************************************************
* Function Definitions
*******************************************************************************/

/*
 * @brief BLE application init
 * @param None
 * @retval 0: success, -1: failed
 */
int ble_app_init(void)
{
    if(ble_init(&ble_callback) != 0)
    {
        LOG_ERR("BLE init failed");
        return -1;
    }

    ble_custom_service_init(&ble_custom_gatt_cb);

    if (ble_adv_start() != 0)
    {
        LOG_ERR("BLE adv start failed");
        return -1;
    }
    return 0;
}


/*
 * @brief: Send sensor data via BLE
 * 
 * @param p_data: pointer to the data to send 
 * @param p_length: pointer to the length of the data to send
 * @param frame_cnt: frame count
 * @return int: 0 on success or negative code otherwise
 */
int sensor_data_send_ble(uint8_t* p_data, uint16_t* p_length, uint32_t frame_cnt)
{
    __ASSERT_NO_MSG(p_data != NULL);
    __ASSERT_NO_MSG(p_length != NULL);
    uint16_t input_len = *p_length;
#if (SENSOR_DATA_SEND_BOARDCAST != 0)
    uint8_t send_payload[1 + sizeof(frame_cnt) + input_len]; // 1B frame heaader + 4B frame count + 12B sensor data
    memset(send_payload, 0, sizeof(send_payload));
    send_payload[0] = SENSOR_DATA_HEADER;
    memcpy(send_payload + 1, &frame_cnt, sizeof(frame_cnt));
    memcpy(send_payload + 1 +sizeof(frame_cnt), p_data, input_len);
    return ble_set_custom_adv_payload(send_payload, 1 + sizeof(frame_cnt) + input_len);
#else /* !(SENSOR_DATA_SEND_BOARDCAST != 0) */
    uint8_t send_payload[sizeof(frame_cnt) + input_len]; //4B frame count + 12B sensor data
    memset(send_payload, 0, sizeof(send_payload));
    memcpy(send_payload, &frame_cnt, sizeof(frame_cnt));
    memcpy(send_payload +sizeof(frame_cnt), p_data, input_len);
    int send_status = sensor_char_send_indication(send_payload, sizeof(frame_cnt) + input_len);
    if(send_status != 0)
    {
        LOG_ERR("BLE send failed with status %d", send_status);
        return send_status;
    }
    return 0;
#endif /* End of (SENSOR_DATA_SEND_BOARDCAST != 0) */
    
}

int main(void)
{
    LOG_INF("Firmware version: %d.%d.%d", FW_MAJOR_VERSION, FW_MINOR_VERSION, FW_BUIILD_VERSION);
	if (ble_app_init() != 0)
    {
        LOG_ERR("App BLE init failed");
        return -1;
    }

#if (CONFIG_BOARD_XIAO_BLE == 1) 
    if (sensor_init() != 0)
    {
        LOG_ERR("IMU sensor init failed");
        return -1;
    }
#endif /* End of (CONFIG_BOARD_XIAO_BLE == 1)  */
    while(1)
    {
        if(is_ble_connect == false)
        {
            k_sleep(K_MSEC(1000));
            LOG_WRN("Waiting for BLE connection");
            continue;
        }
        static uint32_t frame_cnt = 0;
        frame_cnt++;
        uint8_t sensor_data[50] = {0};
        uint16_t sensor_data_len = 0;
#if (CONFIG_BOARD_XIAO_BLE == 1) 

        if (sensor_sampling(sensor_data, &sensor_data_len) != 0)
        {
            LOG_ERR("Sensor sampling failed");
        }
        if(sensor_data_len == SENSOR_PACKET_LEN)
        {
            if (sensor_data_send_ble(sensor_data, &sensor_data_len, frame_cnt) != 0)
            {
                LOG_ERR("Sensor data send BLE failed");
            }
            else
            {
                LOG_INF("Frame %d sent", frame_cnt);
                LOG_DBG("Accel (m/s^2): %.02f, %.02f, %.02f \r\n", *(float*)sensor_data, *(float*)(sensor_data + 4), *(float*)(sensor_data + 8));
                LOG_DBG("Gyro (radians/s): %.02f, %.02f, %.02f \r\n", *(float*)(sensor_data + 12), *(float*)(sensor_data + 16), *(float*)(sensor_data + 20));
            }
        }
        else
        {
            LOG_WRN("Sensor data length %d is invalid", sensor_data_len);
        }
#else // (CONFIG_BOARD_XIAO_BLE != 1) 
        // Simulate sending sensor data
        for(uint8_t count=0; count< SENSOR_PACKET_LEN; count++)
        {
            sensor_data[count] = count;
        }
        sensor_data_len = SENSOR_PACKET_LEN;
        if (sensor_data_send_ble(sensor_data, &sensor_data_len, frame_cnt) != 0)
        {
            LOG_ERR("Sensor data send BLE failed");
        }
        else
        {
            LOG_INF("Frame %d sent", frame_cnt);
        }
#endif /* End of (CONFIG_BOARD_XIAO_BLE == 1)  */
        k_sleep(K_MSEC(SAMPLING_RATE_MS));
    }
	return 0;
}

src/imu_driver.c
/*******************************************************************************
* Title                 :   Driver for LSM6DSL IMU
* Filename              :   imu_driver.c
* Origin Date           :   2023/10/09
* Version               :   0.0.0
* Compiler              :   nRF connect SDK v2.4
* Target                :   nRF52840 SoC
* Notes                 :   None
*******************************************************************************/


/** \file imu_driver.c
 *  \brief This module contains the
 */
/******************************************************************************
* Includes
*******************************************************************************/

/******************************************************************************
* Module Preprocessor Constants
*******************************************************************************/


/******************************************************************************
* Module Preprocessor Macros
*******************************************************************************/
#include "zephyr/logging/log.h"
#define MODULE_NAME			        imu_sensor
#define MODULE_LOG_LEVEL	        LOG_LEVEL_DBG
LOG_MODULE_REGISTER(MODULE_NAME, MODULE_LOG_LEVEL);

/******************************************************************************
* Module Typedefs
*******************************************************************************/

/******************************************************************************
* Module Variable Definitions
*******************************************************************************/

/******************************************************************************
* Function Prototypes
*******************************************************************************/

/******************************************************************************
* Function Definitions
*******************************************************************************/
#if (CONFIG_BOARD_XIAO_BLE == 1) 

#if (CONFIG_DT_HAS_ST_LSM6DSL_ENABLED != 0)
#include <zephyr/device.h>
#include <zephyr/drivers/sensor.h>
#else /* !(CONFIG_DT_HAS_ST_LSM6DSL_ENABLED != 0) */
#error "No enabled LSM6DSL sensor node"
#endif /* End of (CONFIG_DT_HAS_ST_LSM6DSL_ENABLED != 0) */

const struct device *const lsm6dsl_dev = DEVICE_DT_GET_ONE(st_lsm6dsl);
// Possible ODR of LSM6DSL

int sensor_init()
{
    if (!device_is_ready(lsm6dsl_dev)) 
    {
		LOG_ERR("LSM6DSL device not ready");
        return -1;
	}

    if (sensor_sample_fetch(lsm6dsl_dev) < 0) 
    {
        LOG_ERR("Cannot fetch sample");
        return -1;
	}
    return 0;
}

/*
 * @brief: Convert sensor data to float
 * 
 * @param val: input sensor data
 * @return float 
 */
float sensor_data_convert(struct sensor_value *val)
{
    __ASSERT_NO_MSG(val != NULL);
    return (float)(val->val1 + val->val2 / 1000000.0f);
}

/*
 * @brief: Sensor sampling
 * 
 * @param (out) p_data: pointer to store new data
 * @param (out) p_length: pointer to store length of new data
 * @return int: 0 on success or negative code otherwise
 */
int sensor_sampling(uint8_t* p_data, uint16_t* p_length)
{
    int status = 0;
    __ASSERT_NO_MSG(p_data != NULL);
    __ASSERT_NO_MSG(p_length != NULL);

    static struct sensor_value accel_x, accel_y, accel_z;
    static struct sensor_value gyro_x, gyro_y, gyro_z;

    static float accel_x_f, accel_y_f, accel_z_f;
    static float gyro_x_f, gyro_y_f, gyro_z_f;
    
    if (sensor_sample_fetch_chan(lsm6dsl_dev, SENSOR_CHAN_ACCEL_XYZ) < 0) {
        LOG_ERR("Cannot fetch accel sample");
        return -1;
    }
    if (sensor_channel_get(lsm6dsl_dev, SENSOR_CHAN_ACCEL_X, &accel_x) < 0) {
        LOG_ERR("Cannot get accel_x data");
        return -1;
    }
    if (sensor_channel_get(lsm6dsl_dev, SENSOR_CHAN_ACCEL_Y, &accel_y) < 0) {
        LOG_ERR("Cannot get accel_y data");
        return -1;
    }
    if (sensor_channel_get(lsm6dsl_dev, SENSOR_CHAN_ACCEL_Z, &accel_z) < 0) {
        LOG_ERR("Cannot get accel_z data");
        return -1;
    }

    /* lsm6dsl gyro */
    if (sensor_sample_fetch_chan(lsm6dsl_dev, SENSOR_CHAN_GYRO_XYZ) < 0) {
        LOG_ERR("Cannot fetch gyro sample");
        return -1;
    }
    if (sensor_channel_get(lsm6dsl_dev, SENSOR_CHAN_GYRO_X, &gyro_x) < 0) {
        LOG_ERR("Cannot get gyro_x data");
        return -1;
    }
    if (sensor_channel_get(lsm6dsl_dev, SENSOR_CHAN_GYRO_Y, &gyro_y) < 0) {
        LOG_ERR("Cannot get gyro_y data");
        return -1;
    }
    if (sensor_channel_get(lsm6dsl_dev, SENSOR_CHAN_GYRO_Z, &gyro_z) < 0) {
        LOG_ERR("Cannot get gyro_z data");
        return -1;
    }

    accel_x_f = sensor_data_convert(&accel_x);
    accel_y_f = sensor_data_convert(&accel_y);
    accel_z_f = sensor_data_convert(&accel_z);

    gyro_x_f = sensor_data_convert(&gyro_x);
    gyro_y_f = sensor_data_convert(&gyro_y);
    gyro_z_f = sensor_data_convert(&gyro_z);

    uint16_t payload_idx = 0;
    memcpy(p_data + payload_idx, &accel_x_f, sizeof(accel_x_f));
    payload_idx += sizeof(accel_x_f);
    memcpy(p_data + payload_idx, &accel_y_f, sizeof(accel_y_f));
    payload_idx += sizeof(accel_y_f);
    memcpy(p_data + payload_idx, &accel_z_f, sizeof(accel_z_f));
    payload_idx += sizeof(accel_z_f);

    memcpy(p_data + payload_idx, &gyro_x_f, sizeof(gyro_x_f));
    payload_idx += sizeof(gyro_x_f);
    memcpy(p_data + payload_idx, &gyro_y_f, sizeof(gyro_y_f));
    payload_idx += sizeof(gyro_y_f);
    memcpy(p_data + payload_idx, &gyro_z_f, sizeof(gyro_z_f));
    payload_idx += sizeof(gyro_z_f);

    *p_length = payload_idx;
    return status;
}
#endif /* End of (CONFIG_BOARD_XIAO_BLE == 1)  */


src/imu_driver.h
/****************************************************************************
* Title                 :   Driver for LSM6DSL IMU header file
* Filename              :   imu_driver.h
* Origin Date           :   2023/10/09
* Version               :   v0.0.0
* Compiler              :   nRF connect SDK v2.4
* Target                :   nRF52840 SoC
* Notes                 :   None
*****************************************************************************/


/** \file imu_driver.h
 *  \brief This module contains .
 *
 *  This is the header file for 
 */
#ifndef IMU_DRIVER_H_
#define IMU_DRIVER_H_

/******************************************************************************
* Includes
*******************************************************************************/


/******************************************************************************
* Preprocessor Constants
*******************************************************************************/


/******************************************************************************
* Configuration Constants
*******************************************************************************/


/******************************************************************************
* Macros
*******************************************************************************/


/******************************************************************************
* Typedefs
*******************************************************************************/


/******************************************************************************
* Variables
*******************************************************************************/


/******************************************************************************
* Function Prototypes
*******************************************************************************/

#ifdef __cplusplus
extern "C"{
#endif
int sensor_init();
int sensor_sampling(uint8_t* p_data, uint16_t* p_length);
float sensor_data_convert(struct sensor_value *val);
void Function(void);

#ifdef __cplusplus
} // extern "C"
#endif

#endif // IMU_DRIVER_H_

/*** End of File **************************************************************/

src/ble/ble_gatt.h
/****************************************************************************
* Title                 :    header file
* Filename              :   gatt_sensor.h
* Author                :   thuantm5
* Origin Date           :   2023/03/17
* Version               :   v0.0.0
* Compiler              :   nRF connect SDK 2.4
* Target                :   nrf52
* Notes                 :   None
*****************************************************************************/

/*************** INTERFACE CHANGE LIST **************************************
*
*    Date    	Software Version    Initials   	Description
*  2023/03/17    v0.0.0         	thuantm5      Interface Created.
*
*****************************************************************************/

/** \file gatt_sensor.h
 *  \brief This module contains .
 *
 *  This is the header file for 
 */
#ifndef BLE_GATT_SENSOR_H_
#define BLE_GATT_SENSOR_H_

/******************************************************************************
* Includes
*******************************************************************************/
#include <stdint.h>
#include <zephyr/bluetooth/bluetooth.h>
#include <zephyr/bluetooth/uuid.h>
#include <zephyr/bluetooth/gatt.h>

/******************************************************************************
* Preprocessor Constants
*******************************************************************************/
/** @brief UUID of the SERVICE1/ CHAR1 **/
#define SENSOR_SERVICE_UUID         (0xFF00)
#define SENSOR_CHAR_UUID            (0xFF01)
#define COMMAND_CHAR_UUID           (0xFF02)
/******************************************************************************
* Configuration Constants
*******************************************************************************/


/******************************************************************************
* Macros
*******************************************************************************/

/******************************************************************************
* Typedefs
*******************************************************************************/
/* 
 * @brief: GATT read callback
 * @param [out] p_param: (reserved)
 * @para [out] p_len: updated length of p_data
 * @return: pointer to data of the attribute
*/
typedef void* (*ble_gatt_read_cb)(void* p_param, void* p_len); /* Custom GATT READ callbacks */

/* 
 * @brief: custom GATT write callback
 * @param [in] p_data: pointer to data from GATT client
 * @param [in] p_len: pointer that contain receive length from GATT client
 * @return: (reserved)
*/
typedef int (*ble_gatt_write_cb)(void* p_data, void* p_len); /* BLE GATT callbacks custom service*/


typedef struct
{
    ble_gatt_read_cb custom_char_read_cb;    // Application callback for custom char 1 read evt
    ble_gatt_write_cb custom_char_write_cb;    // Application callback for custom char 2 read evt
}ble_custom_gatt_cb_t;

/******************************************************************************
* Variables
*******************************************************************************/


/******************************************************************************
* Function Prototypes
*******************************************************************************/
void ble_custom_service_init(ble_custom_gatt_cb_t* ble_gatt_cb);
int sensor_char_send_indication(uint8_t* p_data, uint16_t len);



#endif // BLE_GATT_SENSOR_H_

/*** End of File **************************************************************/

src/ble/bluetoothle.h
/****************************************************************************
* Title                 :   bluetooth le interface
* Filename              :   bluetoothle.h
* Author                :   Itachi
* Origin Date           :   2022/09/04
* Version               :   v0.0.0
* Compiler              :   nRF connect SDK v2.4
* Target                :   nRF52840 SoC
* Notes                 :   None
*****************************************************************************/

/*************** INTERFACE CHANGE LIST **************************************
*
*    Date    	Software Version    Initials   	Description
*  2022/09/04    v0.0.0         	Itachi      Interface Created.
*
*****************************************************************************/

/** \file bluetoothle.h
 *  \brief This module contains interface for ble with Zephyr RTOS using nRF52840dk
 *
 *  This is the header file for 
 */
#ifndef BLE_BLUETOOTHLE_H_
#define BLE_BLUETOOTHLE_H_

/******************************************************************************
* Includes
*******************************************************************************/
#include <zephyr/bluetooth/bluetooth.h>
#include <zephyr/bluetooth/uuid.h>
#include <zephyr/bluetooth/gatt.h>

#include "ble_gatt.h"
/******************************************************************************
* Preprocessor Constants
*******************************************************************************/


/******************************************************************************
* Configuration Constants
*******************************************************************************/

/******************************************************************************
* Macros
*******************************************************************************/


/******************************************************************************
* Typedefs
*******************************************************************************/
typedef struct
{
    void (*ble_connected_cb)(void);
    void (*ble_disconnected_cb)(void);
    void (*ble_adv_started_cb)(void);
    void (*ble_adv_stopped_cb)(void);
}ble_callback_t;

/******************************************************************************
* Variables
*******************************************************************************/

/******************************************************************************
* Function Prototypes
*******************************************************************************/
int ble_init(ble_callback_t *cb);
int ble_adv_start(void);
int ble_adv_stop(void);
int ble_set_adv_name(char* p_name);
int ble_set_custom_adv_payload(uint8_t* p_data, uint8_t len);
#endif // BLE_BLUETOOTHLE_H_

/*** End of File **************************************************************/

src/ble/bluetoothle.c
/*******************************************************************************
* Title                 :   Bluetooth LE 
* Filename              :   bluetoothle.c
* Author                :   Itachi
* Origin Date           :   2022/09/04
* Version               :   0.0.0
* Compiler              :   nRF connect SDK v2.4
* Target                :   nRF52840 SoC
* Notes                 :   None
*******************************************************************************/

/*************** MODULE REVISION LOG ******************************************
*
*    Date       Software Version	Initials	Description
*  2022/09/04       0.0.0	         Itachi      Module Created.
*
*******************************************************************************/

/** \file bluetoothle.c
 * \brief This module contains the
 */
/******************************************************************************
* Includes
*******************************************************************************/
#include "bluetoothle.h"
#include "ble_gatt.h"

/******************************************************************************
* Module configs
*******************************************************************************/
/* Logging relative */
#include <zephyr/logging/log.h>
#define MODULE_NAME			        bluetoothle
#define MODULE_LOG_LEVEL	        LOG_LEVEL_INF
LOG_MODULE_REGISTER(MODULE_NAME, MODULE_LOG_LEVEL);

#define ADV_DEFAULT_DEVICE_NAME     CONFIG_BT_DEVICE_NAME
#define ADV_CUSTOM_DATA_TYPE        BT_DATA_MANUFACTURER_DATA
#define ADV_PACKET_MAX_LEN          (31)
#define ADV_CUSTOM_PAYLOAD_LEN      (0) // Max length - 1B type - 1B length
#define ADV_NAME_MAX_LEN            (ADV_PACKET_MAX_LEN - (ADV_CUSTOM_PAYLOAD_LEN + 2) - 2)


#define CONF_ADV_NAME_APPEND_MAC_ADDR   (1) // 1: include mac address in adv name, 0: not include

/******************************************************************************
* Module Typedefs
*******************************************************************************/

/******************************************************************************
* Module Variable Definitions
*******************************************************************************/
volatile static bool is_advertising=false;
static struct bt_conn *current_conn;

#if (ADV_NAME_MAX_LEN < 0)
#warning "ADV_NAME_MAX_LEN should be greater than 0"
#else /* !(ADV_NAME_MAX_LEN < 0) */
static char ADV_NAME[ADV_NAME_MAX_LEN] = ADV_DEFAULT_DEVICE_NAME;
#endif /* End of (ADV_NAME_MAX_LEN < 0) */

uint8_t ADV_CUSTOM_PAYLOAD[ADV_CUSTOM_PAYLOAD_LEN] = {0};

static struct bt_data ADV_DATA[] = 
{
    BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),                   /* General discoverable mode */
#if (ADV_NAME_MAX_LEN < 0)
#warning "ADV_NAME_MAX_LEN is negative"
#else
    BT_DATA(BT_DATA_NAME_COMPLETE, ADV_NAME, sizeof(ADV_NAME)-1), /* Device name */
#endif

#if !(ADV_CUSTOM_PAYLOAD_LEN <= 0)
    BT_DATA(ADV_CUSTOM_DATA_TYPE, ADV_CUSTOM_PAYLOAD, ADV_CUSTOM_PAYLOAD_LEN)  /* Custom payload */
#endif
};

/* Bluetooth applicatiton callbacks */
static ble_callback_t ble_cb_app = {
    .ble_connected_cb = NULL,
    .ble_disconnected_cb = NULL,
    .ble_adv_started_cb = NULL,
    .ble_adv_stopped_cb = NULL,
};

/******************************************************************************
* Static Function Definitions
*******************************************************************************/
static void MTU_exchange_cb(struct bt_conn *conn, uint8_t err, struct bt_gatt_exchange_params *params)
{
	if (!err) 
    {
		LOG_INF("MTU exchange done. "); 
        LOG_INF("MTU: %d", bt_gatt_get_mtu(current_conn) - 3);
	} else 
    {
		LOG_ERR("MTU exchange failed (err %" PRIu8 ")", err);
	}
}

static void request_mtu_exchange(void)
{	int err;
	static struct bt_gatt_exchange_params exchange_params;
	exchange_params.func = MTU_exchange_cb;

	err = bt_gatt_exchange_mtu(current_conn, &exchange_params);
	if (err)
    {
		LOG_ERR("MTU exchange failed (err %d)", err);
	}  else 
    {
		LOG_INF("MTU exchange pending");
	}
}

static void request_data_len_update(void)
{
	int err;
	err = bt_conn_le_data_len_update(current_conn, BT_LE_DATA_LEN_PARAM_MAX);
    if (err) 
    {
        LOG_ERR("LE data length update request failed: %d",  err);
    }
}

static bool ble_is_advertising(void)
{
    return is_advertising;
}

static void ble_set_advertising(bool is_adv)
{
    is_advertising = is_adv;
}

static void on_ble_connect(struct bt_conn *conn, uint8_t err)
{
	if(err) 
    {
		LOG_ERR("BLE connection err: %d, re-advertising \n", err);
		return;
	}
    current_conn= bt_conn_ref(conn); 
    LOG_INF("BLE Connected");
    LOG_INF("Initializing data length update and MTU exchange");
	request_mtu_exchange();
	request_data_len_update();

    if(ble_cb_app.ble_connected_cb != NULL)
    {
        ble_cb_app.ble_connected_cb();
    }

    struct bt_conn_info info;
    err = bt_conn_get_info(conn, &info);
    if (err) {
        LOG_ERR("bt_conn_get_info() returned %d", err);
        return;
    }
    double connection_interval = info.le.interval*1.25; // in ms
    uint16_t supervision_timeout = info.le.timeout*10; // in ms
    LOG_INF("Connection parameters: interval %.2f ms, latency %d intervals, timeout %d ms", connection_interval, info.le.latency, supervision_timeout);
}

static void on_ble_disconnect(struct bt_conn *conn, uint8_t reason)
{
	LOG_INF("BLE Disconnected (reason: %d)", reason);
    if (current_conn) {
		bt_conn_unref(current_conn);
		current_conn = NULL;
	}
    if(ble_cb_app.ble_disconnected_cb != NULL)
    {
        ble_cb_app.ble_disconnected_cb();
    }
}

void on_ble_param_updated(struct bt_conn *conn, uint16_t interval, uint16_t latency, uint16_t timeout)
{
    LOG_INF("BLE param updated");
    // Dump all params
    LOG_INF("BLE interval: %.02f ms", (float)interval * 1.25);
    LOG_INF("BLE timeout: %.02f ms", (float)timeout * 10);
    LOG_INF("BLE latency: %d", latency);
}

void on_ble_data_len_updated(struct bt_conn *conn, struct bt_conn_le_data_len_info *info)
{
    LOG_INF("BLE data len updated");
    // Dump all params
    LOG_INF("BLE tx_max_len: %dB", info->tx_max_len);
    LOG_INF("BLE tx_max_time: %d (us)", info->tx_max_time);
    LOG_INF("BLE rx_max_len: %dB", info->rx_max_len);
    LOG_INF("BLE rx_max_time: %d (us)", info->rx_max_time);
}

/******************************************************************************
* Function Definitions
*******************************************************************************/
int ble_adv_start(void)
{
    if(ble_is_advertising())
    {
        LOG_WRN("Advertising already started\n");
        return 0;
    }
	int errorcode = bt_le_adv_start(BT_LE_ADV_CONN, ADV_DATA, ARRAY_SIZE(ADV_DATA), NULL, 0);
    if (errorcode) {
        LOG_ERR("Couldn't start advertising (err = %d)", errorcode);
        return errorcode;
    }
    ble_set_advertising(true);
    LOG_INF("Advertising successfully started\n");
    if(ble_cb_app.ble_adv_started_cb != NULL)
    {
        ble_cb_app.ble_adv_started_cb();
    }
    return 0;
}

int ble_adv_stop(void)
{
    if(!ble_is_advertising())
    {
        LOG_WRN("Advertising already stopped\n");
        return 0;
    }
    int errorcode = bt_le_adv_stop();
    if (errorcode) {
        LOG_ERR("Couldn't stop advertising (err = %d)", errorcode);
        return errorcode;
    }
    ble_set_advertising(false);
    LOG_INF("Advertising successfully stopped\n");
    if(ble_cb_app.ble_adv_stopped_cb != NULL)
    {
        ble_cb_app.ble_adv_stopped_cb();
    }
    return 0;
}

int ble_init(ble_callback_t* p_app_cb)
{
    int errorcode= 0;

    /* Bluetooth zephyr internal callbacks */
    static struct bt_conn_cb ble_cb = {
        .connected 		= &on_ble_connect,
        .disconnected 	= &on_ble_disconnect,
        .le_param_updated = &on_ble_param_updated,
        .le_data_len_updated = &on_ble_data_len_updated,
    };

    LOG_INF("BLE initializing \n\r");

    /* Initialize the application callbacks */
    if(p_app_cb != NULL)
    {
        ble_cb_app = *p_app_cb;
    }
    /* Assign callbacks for connection events */
    bt_conn_cb_register(&ble_cb);

    /* BLE initialization */
    errorcode = bt_enable(NULL);
    if(errorcode)
    {
        LOG_ERR("bt_enable return err %d \r\n", errorcode);
        return errorcode;
    }
    LOG_INF("BLE init succesfully");

#if (CONF_ADV_NAME_APPEND_MAC_ADDR != 0)
    // Get the BLE address base on NRF_FICR->DEVICEADDR
    char adv_device_name[ADV_NAME_MAX_LEN] = {0};
    // Append 4 last digits of mac address to device name 
    sprintf(adv_device_name, "%s%02X%02X%02X%02X%02X%02X", ADV_DEFAULT_DEVICE_NAME,
                    (uint8_t)( (NRF_FICR->DEVICEADDR[1] >> 8) & 0xFF), (uint8_t)(NRF_FICR->DEVICEADDR[1] & 0xFF),
                    (uint8_t)( (NRF_FICR->DEVICEADDR[0] >> 24) & 0xFF), (uint8_t)( (NRF_FICR->DEVICEADDR[0] >> 16) & 0xFF),
                    (uint8_t)( (NRF_FICR->DEVICEADDR[0] >> 8) & 0xFF), (uint8_t)(NRF_FICR->DEVICEADDR[0] & 0xFF));
    ble_set_adv_name(adv_device_name);
#endif // End of CONF_ADV_NAME_APPEND_MAC_ADDR

    return 0;
}

int ble_set_adv_name(char* p_name)
{
    __ASSERT_NO_MSG(p_name != NULL);
    if(strlen(p_name) > ADV_NAME_MAX_LEN)
    {
        LOG_ERR("BLE name too long, %d/%d", strlen(p_name), ADV_NAME_MAX_LEN);
    }
    // Find index of BT_DATA_NAME_COMPLETE in ADV_DATA[]
    for(int index = 0; index < ARRAY_SIZE(ADV_DATA); index++)
    {
        if(ADV_DATA[index].type == BT_DATA_NAME_COMPLETE)
        {
            memset((void *)ADV_DATA[index].data, 0, ADV_DATA[index].data_len);
            ADV_DATA[index].data_len = strlen(p_name);
            memcpy((void *)ADV_DATA[index].data, p_name, strlen(p_name));
            LOG_INF("BLE name set to %s", p_name);
            if(ble_is_advertising())
            {
                return bt_le_adv_update_data(ADV_DATA, ARRAY_SIZE(ADV_DATA), NULL, 0);
            }
            else
            {
                return 0;
            }
        }
    }
    LOG_ERR("Couldn't find BT_DATA_NAME_COMPLETE in ADV_DATA");
    return -1;
}

/* 
 * @brief: Set the custom payload of BLE advertising packet with given ADV type is ADV_CUSTOM_DATA_TYPE
 * 
 * @param p_data: pointer to the data
 * @param len: length of the data
 * @return int: 0 on success or negative code otherwise
 */
int ble_set_custom_adv_payload(uint8_t* p_data, uint8_t len)
{   
    __ASSERT_NO_MSG(p_data != NULL);
    if(len > ADV_CUSTOM_PAYLOAD_LEN)
    {
        LOG_ERR("BLE custom payload too long, max length is %d data will be truncated", ADV_CUSTOM_PAYLOAD_LEN);
    }
    // Find index of ADV_CUSTOM_DATA_TYPE in ADV_DATA[]
    for(int index = 0; index < ARRAY_SIZE(ADV_DATA); index++)
    {
        if(ADV_DATA[index].type == ADV_CUSTOM_DATA_TYPE)
        {
            memset((void *)ADV_DATA[index].data, 0, ADV_DATA[index].data_len);
            ADV_DATA[index].data_len = len;
            memcpy((void *)ADV_DATA[index].data, p_data, len);
            if(ble_is_advertising())
            {
                return bt_le_adv_update_data(ADV_DATA, ARRAY_SIZE(ADV_DATA), NULL, 0);
            }
            else
            {
                return 0;
            }
        }
    }
    LOG_ERR("Couldn't find ADV_CUSTOM_DATA_TYPE in ADV_DATA");
    return -1;
}